diag_manager/fms_diag_file_object.F90

```diff
@@ -752,51 +779,182 @@ subroutine add_axes(this, axis_ids, diag_axis, naxis, yaml_id, buffer_id, output
+subroutine update_write_on_this_pe(this, write_on_this_pe)
+  class(fmsDiagFile_type),                 intent(inout) :: this             !< The file object
+  logical,                                 intent(in)    :: write_on_this_pe !< .True. if the current PE is in
+                                                                             !! subregion
+
+  select type (this)
+  type is (subRegionalFile_type)
+    if (this%write_on_this_pe) this%write_on_this_pe = write_on_this_pe
+  end select
+end subroutine update_write_on_this_pe
+
+!> @brief Query for the write_on_this_pe member of the diag file object
+!! @return the write_on_this_pe member of the diag file object
+function get_write_on_this_pe(this) &
+  result(rslt)
+  class(fmsDiagFile_type),                 intent(inout) :: this             !< The file object
+  logical :: rslt
+  rslt = .true.
+  select type (this)
+  type is (subRegionalFile_type)
+    rslt= this%write_on_this_pe
+  end select
+end function get_write_on_this_pe
+
+!< @brief Determine if an axis is already in the list of axis for a diag file
+!! @return .True. if the axis is already in the list of axis for a diag file
+function does_axis_exist(this, var_axis_id) &
+  result(rslt)
+  class(fmsDiagFile_type), intent(inout) :: this         !< The file object
+  integer,                 intent(in)    :: var_axis_id  !< Variable axis id to check
+
+  logical :: rslt
+  integer :: j !< For do loops
+
+  rslt = .false.
+  do j = 1, this%number_of_axis
+    !> Check if the axis already exists, move on
+    if (var_axis_id .eq. this%axis_ids(j)) then
+      rslt = .true.
+      return
+    endif
+  enddo
+end function
+
+!> @brief Define a new sub axis
+subroutine define_new_subaxis(this, var_axis_ids, x_y_axis_id, is_cube_sphere, diag_axis, naxis)
+  class(fmsDiagFile_type),                 intent(inout) :: this              !< The file object
+  integer,                                 INTENT(inout) :: var_axis_ids(:)   !< Original variable axis ids
+  integer,                                 INTENT(in)    :: x_y_axis_id(:)    !< The ids of the x and y axis
+  logical,                                 intent(in)    :: is_cube_sphere    !< .True. if the axis is in the cubesphere
+  integer,                                 intent(inout) :: naxis             !< Number of axis current registered
+  class(fmsDiagAxisContainer_type),        intent(inout) :: diag_axis(:)      !< Diag_axis object
+
+  logical :: write_on_this_pe !< .True. if the current PE is in the subregion
+  integer :: i, j             !< For do loop
+
+  select case (this%get_file_sub_region_grid_type())
+  case(latlon_gridtype)
+    call define_new_subaxis_latlon(diag_axis, x_y_axis_id, naxis, this%get_file_sub_region(), is_cube_sphere, &
+      write_on_this_pe)
+    call this%update_write_on_this_pe(write_on_this_pe)
+    if (.not. this%get_write_on_this_pe()) return
+    call this%add_new_axis(naxis)
+    call this%add_new_axis(naxis-1)
```

Comment:
could this potentially create duplicate axis ids?

Reply:
No, the code only goes into `define_new_subaxis` if the subaxis have not been created for the file. 
`define_new_subaxis_latlon` is going to create new subaxis (one for x and y), that is why there is a `naxis` and `naxis-1`